### PR TITLE
BUG: Look up methods on MaskedArray in _frommethod

### LINKED ISSUE #8019: np.ma.count and np.ma.copy on list input
The `np.ma.count` and `np.ma.copy` behave strangly if the input is a `list` or `tuple` (probably also other python builtins).

```
>>> np.ma.copy([1,2,3])  # unexpected behaviour
[1, 2, 3]

>>> np.copy([1,2,3])  # expected behaviour
array([1, 2, 3])

>>> np.ma.count([1,2,3])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Programming\Anaconda\envs\test\lib\site-packages\numpy\ma\core.py", line 6389, in __call__
    return method(*args, **params)
TypeError: count() takes exactly one argument (0 given)
```

The reason for this behaviour is that `list` **has** a `count` and `copy` method and `np.ma._frommethod` tries to call it instead of `np.ma.MaskedArray.count` (and copy).

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #8019

Squashed version of #8060, with the addition of the suggestions I made there in the comments

### CODE REVIEW

**Comment by MSeifert04** [comment]
It would be great if #8019 is finally adressed. :) Thanks for working on it.

**Comment by eric-wieser** [comment]
@MSeifert04: I'm actually targeting this because this can remove a lot of special-casing in `np.ma` functions.

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: Why look up the method on the class rather than just `getattr(marr, method_name, None)`?

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: One could change `MaskedArray.__new__`, though perhaps calling a class constructor and not getting back a new instance might also be surprising... Perhaps best might have been to have had a logical distinction between the `np.ma.MaskedArray` class and a `np.ma.masked_array`/`np.ma.array` function...

Overall, this is probably the best solution.

File: numpy/ma/core.py
Comment (eric-wieser) [inline numpy/ma/core.py]: Seems tidier I guess (and closed to how dunder lookup works), and makes it work on masked record arrays.

You've actually made me notice a bug here, in that `reversed` should only apply to the method call. Either way, that bug was here all along, and can wait for another PR.

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: I think with your other change it is now safe (and shorter though a bit slower) to just do 
```
compressed = _frommethod('compressed')
```

File: numpy/ma/core.py
Comment (eric-wieser) [inline numpy/ma/core.py]: > One could change `MaskedArray.__new__`

This is the correct approach, and already how `ndarray.__new__` behaves. Now take a look at `MaskedArray.__new__`, and tell me how easy that changes looks. That's why I'm calling this a workaround!

File: numpy/ma/core.py
Comment (eric-wieser) [inline numpy/ma/core.py]: I agree. Until just now, I thought that this was only for methods that exist in `np`, but it is not

**Comment by mhvk** [comment]
Looks good!

File: numpy/ma/core.py
Comment (eric-wieser) [inline numpy/ma/core.py]: I'll add another commit for this that fixes a bunch of other methods as well, that were out of scope for the original PR

Comment (eric-wieser) [inline numpy/ma/core.py]: @mhvk: actually, can I bump that method cleanup to another PR?

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: It does look really weird to me to look it up on the class! But of course if one looks up on `marr`, one cannot pass `marr` as first argument below. If I were to write this, I'd probably do:
```
method = getattr(marr, 'method_name', None)
if method is not None:
    return method(*args, **params)
else:
    return getattr(np, method_name)(marr, *args, **params)
```
Anyway, no big deal either way.

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: Yes, better in another PR. As is, this is a bug fix, but that would be maintenance.

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: I think one would add right after the [first line](https://github.com/numpy/numpy/blob/master/numpy/ma/core.py#L2766):
```
if _data is data and (type(_data) is cls or subok and isinstance(_data, cls)):
    return _data
```

Comment (mhvk) [inline numpy/ma/core.py]: Just to be sure: I think for a bug fix you're current approach is preferable!

Comment (eric-wieser) [inline numpy/ma/core.py]: @mhkv: Not enough. I think you also need to check that all of the other arguments are their default value

Comment (mhvk) [inline numpy/ma/core.py]: Yes, makes sense.

File: numpy/ma/core.py
Comment (mhvk) [inline numpy/ma/core.py]: In another PR is slightly better, but really this is making work, so whatever you feel like!

Comment (eric-wieser) [inline numpy/ma/core.py]: I'd rather this just be merged as is, to be honest

Comment (mhvk) [inline numpy/ma/core.py]: No problem.

**Comment by mhvk** [comment]
This one should go in 12.1, I think, since it was a specific, somewhat long-standing issue that got solved.

Since this was reviewed before, I'm quite happy to merge once the tests pass.

**Comment by eric-wieser** [comment]
@mhvk: tests have passed, but travis isn't pinging github correctly

**Review by ahaldane — APPROVED** [review]
It checks out to me. We can merge once the testing system is happy.

**Comment by ahaldane** [comment]
let's see if that gets the tests to register... if not I think it's OK merge anyway since the tests do pass.

**Comment by ahaldane** [comment]
OK, merging. Thanks @eric-wieser !

**Comment by charris** [comment]
@eric-wieser @mhvk Was it intended to backport this?

**Comment by mhvk** [comment]
@charris - yes, this was meant to be backported to 1.12.1; I'll try to do the `backport-candidate` labelling in the future.

**Comment by charris** [comment]
No, don't. I'll just cherry-pick it into #8739.

**Comment by charris** [comment]
Oops, I thought you were volunteering to do the backport...

**Pull request merged**
